ballista/rust/scheduler/src/state/shuffle_reaper.rs

```diff
@@ -0,0 +1,52 @@
+use ballista_core::client::BallistaClient;
+use ballista_core::serde::scheduler::PartitionLocation;
+use log::error;
+use std::collections::HashMap;
+use tokio::sync::mpsc::Receiver;
+
+pub struct ShuffleReaper {
+    receiver: Receiver<Vec<PartitionLocation>>,
+}
+
+impl ShuffleReaper {
+    pub fn new(receiver: Receiver<Vec<PartitionLocation>>) -> Self {
+        Self { receiver }
+    }
+
+    pub fn start(self) {
+        tokio::task::spawn(async move {
+            let mut clients: HashMap<String, BallistaClient> = Default::default();
+            let mut receiver = self.receiver;
+            while let Some(locations) = receiver.recv().await {
+                for location in locations {
+                    if let Some(client) = clients.get_mut(&location.executor_meta.id) {
```

Comment:
I wonder if there could be some kind of 'get or insert' pattern utilised for map here. Found out about entry api (https://doc.rust-lang.org/stable/std/collections/hash_map/struct.HashMap.html#method.entry), but best I could come up with was this which has a downside that executor_id gets always cloned.
```rust
                    let executor_id = location.executor_meta.id.clone();
                    if let Ok(client) = match clients.entry(executor_id) {
                        Entry::Occupied(entry) => Ok(entry.into_mut()),
                        Entry::Vacant(entry) => BallistaClient::try_new(
                            &location.executor_meta.host,
                            location.executor_meta.grpc_port,
                        )
                        .await
                        .map(|client| entry.insert(client)),
                    } {
                        client
                            .delete_partition(&location.path)
                            .await
                            .unwrap_or_else(|e| {
                                error!(
                                "Error deleting partition {:?} from executor {:?}: {:?}",
                                location.path, location.executor_meta.id, e
                            )
                            });
                    } else {
                        error!(
                            "Failed to connect to executor {}",
                            location.executor_meta.id
                        );
                    }
```

Reply:
yeah, this seems more convoluted to me.